### ISSUE TITLE
Generic type `T` inferred as `object` even though implementation returns `?object`

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#2008]
```php
<?php

function returnsObjectOrNull() : ?object
{
    return random_int(0, 1) ? (object) [] : null;
}

/**
 * @psalm-template T as object
 * @psalm-param callable() : T $callback
 * @psalm-return T
 */
function makeResultSet(callable $callback)
{
    return $callback();
}

echo get_class(makeResultSet('returnsObjectOrNull'));
```

https://psalm.dev/r/91a6f0e94c

The code above should report invalid usage of `get_class()`, since `?object` may be returned by `makeResultSet` in this scenario. 

Possibly similar to #1999

**Comment by Ocramius** [comment]
Similar example, without `get_class()` weirdness (WHY DOES `get_class` ALLOW `null`, PHP?!?!?!?!):

```php
<?php
  
class HelloWorld
{
    public function sayHello() : void
    {
        echo 'hello world';
    }
}


function returnsObjectOrNull() : ?HelloWorld
{
    return random_int(0, 1) ? new HelloWorld() : null;
}

/**
 * @psalm-template T as object
 * @psalm-param callable() : T $callback
 * @psalm-return T
 */
function makeResultSet(callable $callback)
{
    return $callback();
}

makeResultSet('returnsObjectOrNull')
  ->sayHello();
```

https://psalm.dev/r/990d073939

**Comment by muglug** [comment]
So this is not strictly a bug - at least, using file-namespaced functions was not designed to work. Using static methods does: https://psalm.dev/r/a1dd9c2f77 generates an error (because the types don’t match)

**Comment by muglug** [comment]
Do you think it needs to work with functions that aren't globally available?

**Comment by Ocramius** [comment]
>  functions that aren't globally available?

Could you clarify on what this means?

**Comment by muglug** [comment]
Globally-available functions are both those autoloaded by Composer (via `autoload_files`) and builtin functions e.g. `strlen`.

**Comment by Ocramius** [comment]
Hmm, the above are part of the file: is that a limitation of the tooling?

If this works with globally available functions, that's fine IMO

**Comment by muglug** [comment]
> Hmm, the above are part of the file: is that a limitation of the tooling?

Yeah, at the moment only static methods and globally-available functions are supported by the genericised callable return type extraction mechanism.
